lib/pedicel/validator.rb

```diff
@@ -1,70 +1,106 @@
 require 'dry-validation'
 require 'base64'
+require 'openssl'
 
 module Pedicel
+  # Validation class for Apple Pay Payment Token and associated data:
+  # https://developer.apple.com/library/content/documentation/PassKit/Reference/PaymentTokenJSON/PaymentTokenJSON.html
+  # This purposefully only does syntactic validation (as opposed to semantic).
   class Validator
     class Error < StandardError; end
     class TokenFormatError < Error; end
     class TokenDataFormatError < Error; end
 
-    DRY_CUSTOM_PREDICATE_ERRORS = {
-      base64?: 'invalid base64',
-      hex?: 'invalid hex',
-      pan?: 'invalid pan',
-      yymmdd?: 'invalid date format YYMMDD',
-    }
-
     module Predicates
       include Dry::Logic::Predicates
 
-      predicate(:base64?) { |value| !Base64.decode64(value).nil? rescue false }
+      CUSTOM_PREDICATE_ERRORS = {
+        base64?:          'invalid base64',
+        hex?:             'invalid hex',
+        pan?:             'invalid pan',
+        yymmdd?:          'invalid date format YYMMDD',
+        ec_public_key?:   'is not an EC public key',
+        pkcs7_signature?: 'is not a PKCS7 Signature',
+        eci?:             'not an ECI indicator',
+        hex_sha256?:      'not hex-encoded SHA256',
+        base64_sha256?:   'not base64-encoded SHA256',
+      }.freeze
+
+      match_b = String.new.respond_to?(:match?) ? lambda{|s, re| s.match?(re)} : lambda{|s, re| !!(s =~ re)}
+      # Support Ruby 2.3, but use the faster #match? when available.
+
+      predicate(:base64?) do |x|
+        str?(x) &&
+          match_b.(x, /\A[=A-Za-z0-9+\/]*\z/) && # allowable chars
+          x.length.remainder(4).zero? && # multiple of 4
+          !match_b.(x, /=[^$=]/) && # may only end with ='s
+          !match_b.(x, /===/) # at most 2 ='s
+      end
```

Comment:
Is it intentional that this validation does not allow characters outside the base64 alphabet? I thought e.g. newlines would be acceptable.

Reply:
It is intentional, yes. Is it the right thing to do? I honestly don't know. I'd like to see a bunch of tokens before I decide 🙂

I tried the tighter version because that'd make us aware if something must be loosened; the reverse would require us to remember it, and I don't trust my own memory 😉